lib/wab/impl/configuration.rb

```diff
@@ -1,104 +1,256 @@
+            key = key.to_sym
+            unless node.has_key?(key)
+              ai = key_to_int_ok(path[i + 1])
+              node[key] = ai.nil? ? {} : []
+            end
+            node = node[key]
+          elsif node.is_a?(Array)
+            key = key_to_int(key)
+            if key < node.length && -node.length < key
+              node = node[key]
             else
-              secondary_key = keys.shift
-              if keys.empty?
-                config[primary_key][secondary_key] = value
+              ai = key_to_int_ok(path[i + 1])
+              nn = ai.nil? ? {} : []
+              if key < -node.length
+                node.unshift(nn)
               else
-                tertiary_key = keys.shift
-                if keys.empty?
-                  config[primary_key][secondary_key][tertiary_key] = value
-                end
+                node[key] = nn
               end
+              node = nn
             end
+          else
+            raise WAB::TypeError, "Can not set a member of an #{node.class}."
           end
+        }
+        key = path[-1]
+        if node.is_a?(Hash)
+          key = key.to_sym
+          node[key] = value
+        elsif node.is_a?(Array)
+          key = key_to_int(key)
+          if key < -node.length
+            node.unshift(value)
+          else
+            node[key] = value
+          end
+        else
+          raise WAB::TypeError, "Can not set a member of an #{node.class}."
         end
-        config
+        value
+      end
+
+      def set(path, value)
+        set_map(@map, path, value)
+      end
+      alias []= set
+
+      def get(path)
+        if path.is_a?(Symbol)
+          node = @map[path]
+        else
+          path = path.to_s.split('.') unless path.is_a?(Array)
+          node = @map
+          path.each { |key|
+            if node.is_a?(Hash)
+              node = node[key.to_sym]
+            elsif node.is_a?(Array)
+              i = key.to_i
+              if 0 == i && '0' != key && 0 != key
+                node = nil
+                break
+              end
+              node = node[i]
+            else
+              node = nil
+              break
+            end
+          }
+        end
+        node
+      end
+      alias [] get
+
+      def key_to_int(key)
+        return key if key.is_a?(Integer)
+
+        key = key.to_s if key.is_a?(Symbol)
+        if key.is_a?(String)
+          i = key.to_i
+          return i if i.to_s == key
+        end
+        return key if WAB::Utils.pre_24_fixnum?(key)
+
+        raise WAB::Error, 'path key must be an integer for an Array.'
+      end
+
+      # returns either an int or nil.
+      def key_to_int_ok(key)
+        return key if key.is_a?(Integer)
+
+        key = key.to_s if key.is_a?(Symbol)
+        if key.is_a?(String)
+          i = key.to_i
+          return i if i.to_s == key
+        end
+        return key if WAB::Utils.pre_24_fixnum?(key)
+        nil
       end
```

Comment:
This whole section looks like repeated lines of code to me..
Consider following DRY principle?

Reply:
It is a repeat of the Data class. I figured we would move it to a utilities file but had to run before I could do that.